File: textgen.go
package main

import (
	"crypto/sha1"
	"encoding/hex"
	"flag"
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"strings"
	"sync"
	"time"
)

func main() {
	words, err := ioutil.ReadFile("dictionary.txt")
	if err != nil {
		panic(fmt.Errorf("Missing dictionary.txt - how can I proceed???"))
	}
	var chunkSize int
	var totalSize int
	var intSeed int64
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "%s: Generate random chunks of data\nOptions:\n", os.Args[0])
		flag.PrintDefaults()
	}
	flag.IntVar(&chunkSize, "chunk-size", 10000, "Size of a single file (chunk)")
	flag.IntVar(&totalSize, "total-size", 1000000, "Total size to generate")
	flag.Int64Var(&intSeed, "random-seed", -1, "Random seed")
	flag.Parse()
	if intSeed == -1 {
		intSeed = time.Now().Unix()
	}
	rand.Seed(intSeed)
	wordary := strings.Split(string(words), "\n")
	wordary_len := len(wordary)
	count := totalSize / chunkSize
	var wg sync.WaitGroup
	for count > 0 {
		wg.Add(1)
		go func() {
			defer wg.Done()
			genchunk(wordary, wordary_len, chunkSize)
		}()
		count -= 1
	}
	wg.Wait()
}

func genchunk(source []string, sourcelen, size int) {
	chunk := strings.Builder{}
	chunk.Grow(size)
	f := 0
	for size > f {
		chunk.WriteString(source[rand.Intn(sourcelen)])
		chunk.WriteRune(' ')
		f = chunk.Len()
	}
	chunk.WriteRune('\n')
	s := chunk.String()
	h := sha1.New()
	h.Write([]byte(s))
	fn := hex.EncodeToString(h.Sum(nil))
	dpath := fmt.Sprintf("%s/%s", fn[0:2], fn[1:3])
	fpath := fmt.Sprintf("%s/%s", dpath, fn)
	os.MkdirAll(dpath, 0750)
	ioutil.WriteFile(fpath, []byte(s), 0640)
}

File: textgen.py
"""textgen: generate random chunks of data

Usage: textgen [--chunk-size=<CHUNK_SIZE>] [--total-size=<TOTAL_SIZE>] [--random-seed=<RANDOM_SEED>]

Options:
    --chunk-size=<CHUNK_SIZE>    Size of a chunk of data [default: 10000]
    --total-size=<TOTAL_SIZE>    Total size of stuff to generate [default: 1000000]
    --random-seed=<RANDOM_SEED>  Seed for the RNG. By default will use current time [default: -1]

"""
import docopt
from io import StringIO
from random import seed, choice
import hashlib
from pathlib import Path
import asyncio
from math import ceil
from time import time
async def main(args):
    chunk_size = int(args['--chunk-size'])
    total_size = int(args['--total-size'])
    random_seed = int(args['--random-seed'])
    if random_seed == -1:
        random_seed = int(time())
    seed(random_seed)
    number = ceil(total_size / chunk_size)
    dictionary = None
    try:
        with open('dictionary.txt', 'r', encoding='utf-8') as fh:
            dictionary = [s.strip() for s in fh.readlines()]
    except Exception as e:
        raise RuntimeError(e)
    t = []
    while number > 0:
        number -= 1
        t.append(generate_chunk(dictionary, chunk_size))
    await asyncio.gather(*t)

async def generate_chunk(dictionary, chunk_size):
    s = StringIO()
    while chunk_size > s.tell():
        i = choice(dictionary)
        s.write(i)
        s.write(' ')
    
    s.write("\n")
    r = s.getvalue()
    fn = hashlib.sha1(r.encode('utf-8')).hexdigest()
    dpath = f'res-py/{fn[0:2]}/{fn[1:3]}'
    fpath = f'{dpath}/{fn}'
    Path(dpath).mkdir(parents=True, exist_ok=True)
    with open(fpath, 'w', encoding='utf-8') as fh:
        fh.write(r)


if __name__ == '__main__':
    args = docopt.docopt(__doc__)
    asyncio.run(main(args))

